Lombiq.Hosting.Tenants.Maintenance/Maintenance/ElasticsearchIndices/Startup.cs

```diff
@@ -0,0 +1,162 @@
+using OrchardCore.Modules;
+using OrchardCore.Search.Elasticsearch.Core.Models;
+using OrchardCore.Search.Elasticsearch.Core.Services;
+using System;
+using System.Linq;
+
+namespace Lombiq.Hosting.Tenants.Maintenance.Maintenance.ElasticsearchIndices;
+
+[Feature(FeatureNames.DeleteOrRebuildElasticsearchIndices)]
+public class DeleteElasticsearchIndicesStartup : StartupBase
+{
+    private readonly IShellConfiguration _shellConfiguration;
+
+    public DeleteElasticsearchIndicesStartup(IShellConfiguration shellConfiguration) => _shellConfiguration = shellConfiguration;
+
+    public override void ConfigureServices(IServiceCollection services)
+    {
+        services.BindAndConfigureSection<ElasticsearchIndicesMaintenanceOptions>(
+            _shellConfiguration,
+            "Lombiq_Hosting_Tenants_Maintenance:ElasticsearchIndicesOptions");
+
+        services.AddScoped<IMaintenanceProvider, DeleteElasticsearchIndicesMaintenanceProvider>();
+        services.AddScoped<IMaintenanceProvider, RebuildElasticsearchIndicesMaintenanceProvider>();
+    }
+}
+
+[Feature(FeatureNames.DeleteElasticsearchIndicesBeforeSetup)]
+public class DeleteElasticsearchIndicesBeforeSetupStartup : StartupBase
+{
+    private readonly IShellConfiguration _shellConfiguration;
+    private readonly ShellSettings _shellSettings;
+
+    public DeleteElasticsearchIndicesBeforeSetupStartup(IShellConfiguration shellConfiguration, ShellSettings shellSettings)
+    {
+        _shellConfiguration = shellConfiguration;
+        _shellSettings = shellSettings;
+    }
+
+    public override void ConfigureServices(IServiceCollection services)
+    {
+        services.BindAndConfigureSection<ElasticsearchIndicesMaintenanceOptions>(
+            _shellConfiguration,
+            "Lombiq_Hosting_Tenants_Maintenance:ElasticsearchIndicesOptions");
+
+        // After the setup, the Elasticsearch module can be loaded the regular way.
+        if (!_shellSettings.IsUninitialized()) return;
+
+        // This is necessary because the Elasticsearch module can't be enabled the regular way if this module is added
+        // as a setup feature, otherwise you get a ContentsAdminList shape missing exception on the admin dashboard. For
+        // more info see:
+        var configuration = _shellConfiguration.GetSection("OrchardCore_Elasticsearch");
+        var elasticConfiguration = configuration.Get<ElasticConnectionOptions>();
+        services.Configure<ElasticConnectionOptions>(o => o.ConfigurationExists = true);
+        // Otherwise the ElasticClient won't work. Copied all this from the OC Elasticsearch module.
+#pragma warning disable CA2000 // Call System. IDisposable. Dispose on object created by
+        // 'GetConnectionSettings(elasticConfiguration)' before all references to it are out of scope
+        var settings = GetConnectionSettings(elasticConfiguration);
+#pragma warning restore CA2000
+        services.AddSingleton<IElasticClient>(new ElasticClient(settings));
```

Comment:
A bit of spacing to make this more readable would be nice:

```suggestion
        services.Configure<ElasticConnectionOptions>(o => o.ConfigurationExists = true);
        
        // Otherwise the ElasticClient won't work. Copied all this from the OC Elasticsearch module.
#pragma warning disable CA2000 // Call System. IDisposable. Dispose on object created by
        // 'GetConnectionSettings(elasticConfiguration)' before all references to it are out of scope
        var settings = GetConnectionSettings(elasticConfiguration);
#pragma warning restore CA2000
        
        services.AddSingleton<IElasticClient>(new ElasticClient(settings));
```